sdks/java/core/src/main/java/org/apache/beam/sdk/coders/CoderFactories.java

```diff
@@ -197,92 +169,43 @@ private Method getFactoryMethod(Class<?> coderClazz) {
-      TypeDescriptor<?> coderType = TypeDescriptor.of(coderClazz);
-      TypeDescriptor<T> argumentType = getCodedType(coderType);
-
-      // getInstanceComponents may be implemented in a superclass,
-      // so we search them all for an applicable method. We do not
-      // try to be clever about finding the best overload. It may
-      // be in a generic superclass, erased to accept an Object.
-      // However, subtypes are listed before supertypes (it is a
-      // topological ordering) so probably the best one will be chosen
-      // if there are more than one (which should be rare)
-      for (TypeDescriptor<?> supertype : coderType.getClasses()) {
-        for (Method method : supertype.getRawType().getDeclaredMethods()) {
-          if (method.getName().equals("getInstanceComponents")) {
-            TypeDescriptor<?> formalArgumentType = supertype.getArgumentTypes(method).get(0);
-            if (formalArgumentType.getRawType().isAssignableFrom(argumentType.getRawType())) {
-              return method;
-            }
-          }
-        }
-      }
-
-      throw new IllegalArgumentException(
-          "cannot create a CoderFactory from " + coderType + ": "
-          + "does not have an accessible method "
-          + "'getInstanceComponents'");
-    }
-
-    /**
-     * If {@code coderType} is a subclass of {@link Coder} for a specific
-     * type {@code T}, returns {@code T.class}. Otherwise, raises IllegalArgumentException.
-     */
-    private <T> TypeDescriptor<T> getCodedType(TypeDescriptor<?> coderType) {
-      for (TypeDescriptor<?> ifaceType : coderType.getInterfaces()) {
-        if (ifaceType.getRawType().equals(Coder.class)) {
-          ParameterizedType coderIface = (ParameterizedType) ifaceType.getType();
-          @SuppressWarnings("unchecked")
-          TypeDescriptor<T> token =
-              (TypeDescriptor<T>) TypeDescriptor.of(coderIface.getActualTypeArguments()[0]);
-          return token;
-        }
-      }
-      throw new IllegalArgumentException(
-          "cannot build CoderFactory from class " + coderType
-          + ": does not implement Coder<T> for any T.");
-    }
-
     @Override
     public String toString() {
       return MoreObjects.toStringHelper(getClass())
+          .add("rawType", rawType)
           .add("factoryMethod", factoryMethod)
-          .add("getComponentsMethod", getComponentsMethod)
           .toString();
     }
   }
 
   /**
-   * See {@link #forCoder} for a detailed description of this
-   * {@link CoderFactory}.
+   * See {@link #forCoder} for a detailed description of this {@link CoderFactory}.
    */
   private static class CoderFactoryForCoder<T> implements CoderFactory {
     private final Coder<T> coder;
+    private final TypeDescriptor<T> type;
 
-    public CoderFactoryForCoder(Coder<T> coder) {
+    public CoderFactoryForCoder(TypeDescriptor<T> type, Coder<T> coder){
+      this.type = type;
```

Comment:
Currently we allow a user to say this coder is good for this type descriptor, do we expect coders are only good for a specific type of descriptor and not something else based upon inheritance like I have an AnimalCoder and I want it to handle my Dog type since it will only encode it as an Animal?

Also, what value does getEncodedTypeDescriptor add on a Coder (since it seems like a good candidate to deprecate/remove and internalize the logic to simplify Coders)?
(Only found one usage in ParDo related to state specs and seems to be related to trying to get the type to coder bindings which we should be able to get again or keep from earlier).